File: climbing-stairs/naringst.js
/**
 * @param {number} n
 * @return {number}
 */
/**
 * Note: DP
 * 결국 1개 전의 값에서 1칸을 더 가거나, 2개 전의 값에서 2칸을 더 가는 경우의 수가 생긴다.
 * 1일때에는 1칸을 가는 방법의 경우 뿐이지만, 2일때에는 0칸 가는 것에서 2칸을 가거나 1칸을 가는 것에서 1칸을 더 갈 수 있다.
 * 3일 때에는 1까지 갔을 경우에서 2칸을 더 가거나, 2까지 갔을 경우에서 1칸을 더 가면 되니까 1까지 갔을 경우의 수 + 2까지 갔을 경우의 수
 *
 *
 * Runtime: 41ms, Memory: 48.95MB
 * Time complexity: O(n)
 * Space complexity: O(n)
 *
 */

var climbStairs = function (n) {
  const dp = Array(n).fill(0);
  dp[1] = 1;
  dp[2] = 2;

  for (let i = 3; i < n + 1; i++) {
    dp[i] = dp[i - 2] + dp[i - 1];
  }

  return dp[n];
};

File: longest-substring-without-repeating-characters/naringst.py

#  Runtime: 51ms, Memory: 16.83MB
#  Time complexity: O(len(s))
#  Space complexity: O(len(s))

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        stringArr = []
        maxLength = 0

        for sub in s :
            if sub in stringArr :    
                maxLength = max(maxLength, len(stringArr))
                repeatIdx = stringArr.index(sub)
                stringArr = stringArr[repeatIdx+1 :]

            stringArr.append(sub)
        
        maxLength = max(maxLength, len(stringArr))


        return maxLength
    